plugins/org.python.pydev.core/src/org/python/pydev/core/docutils/ReversePyDocIterator.java

```diff
@@ -0,0 +1,130 @@
+/**
+ * Copyright (c) 2020 by Brainwy Software Ltda
+ */
+package org.python.pydev.core.docutils;
+
+import org.eclipse.jface.text.BadLocationException;
+import org.eclipse.jface.text.IDocument;
+import org.python.pydev.core.IPythonPartitions;
+import org.python.pydev.core.partition.PyPartitionScanner;
+import org.python.pydev.shared_core.partitioner.FastPartitioner;
+import org.python.pydev.shared_core.string.FastStringBuffer;
+
+public class ReversePyDocIterator implements IPyDocIterator {
+
+    private int offset;
+    private IDocument doc;
+
+    private int lastReturned = -1;
+    private FastPartitioner fastPartitioner;
+
+    public ReversePyDocIterator(IDocument doc) throws BadLocationException {
+        this(doc, doc.getLineOfOffset(doc.getLength() - 1));
+    }
+
+    public ReversePyDocIterator(IDocument doc, int startingLine) throws BadLocationException {
+        this.doc = doc;
+        this.offset = doc.getLineOffset(startingLine) + doc.getLineLength(startingLine) - 1;
+        this.fastPartitioner = (FastPartitioner) PyPartitionScanner.checkPartitionScanner(doc);
+    }
+
+    @Override
+    public boolean hasNext() {
+        return offset >= 0;
+    }
+
+    @Override
+    public int getLastReturnedLine() {
+        if (offset < doc.getLength() - 1) {
+            try {
+                lastReturned = doc.getLineOfOffset(offset + 1);
+            } catch (BadLocationException e) {
+                //ignore (keep the last one)
+            }
+        }
+        return lastReturned;
+    }
+
+    /**
+     * @return the next line in the document
+     */
+    @Override
+    public String next() {
+        FastStringBuffer buf = new FastStringBuffer();
+        FastStringBuffer convertedLiteralsBuf = new FastStringBuffer();
+        boolean gotContents = false;
+        boolean gotLiterals = false;
+        try {
+            char ch = doc.getChar(offset);
+            //handle the \r, \n or \r\n
+            if (ch == '\n' || ch == '\r') {
+                offset--;
+                ch = doc.getChar(offset);
+                if (offset >= 0 && ch == '\r') {
+                    offset--;
+                    ch = doc.getChar(offset);
+                }
+            }
+
+            while (offset >= 0) {
+                ch = doc.getChar(offset);
+                if (ch == '\n') {
+                    break;
+                }
+
+                if (fastPartitioner.getContentType(offset) != IPythonPartitions.PY_DEFAULT) {
+                    convertedLiteralsBuf.insert(0, ' ');
+                    gotLiterals = true;
+                } else {
+                    buf.insert(0, ch);
+                    gotContents = true;
+                }
+                offset--;
+            }
+
+            if (gotContents && gotLiterals) {
```

Comment:
This seems too complicated. Why not just add everything to the same `buf` and then just return `buf.toString()`?

Reply:
I made all this handle because if it receives a doc like the testIterator5 example, i.e.:
```
class Foo:
	''' 
	"
	b 
	'''a 
	pass
```
In the line that multiline string ends with some content after, it will return a string like:
`"       a"`

Which will cause problems to the indent verifier in PySelection.getLineThatStartsScope

Reply:
I do think there is a more simplified way to do it, but this is the explanation behind why all these handlers was inserted